File: diving-beet.py
#!/usr/bin/env python

import game
import numpy
import pygame
import time
import sys

beet = game.game(interactive=True)

fullscreen = False

if fullscreen:
    desired_size = None
else:
    desired_size = (1200,800)

pygame.init()
pygame.display.set_caption('Diving Beet')
if desired_size != None:
    screen = pygame.display.set_mode(desired_size)
else:
    screen = pygame.display.set_mode()
width = screen.get_width()
height = screen.get_height()
size = (width,height)
font = pygame.font.Font(None, 36)
pygame.key.set_repeat(100, 100)

def showText(what, where):
    text = font.render(what, 1, (255, 255, 255))
    screen.blit(text, where)

beet_state = beet.new_game(width, height)

def element_name(elem):
    return ''.join(map(chr,list(beet.element_name(elem).get())))

# Initialise the list of insertable elements and their names.
insertable = []
for elem in beet.insertable_elements().get():
    insertable += [(element_name(elem), elem)]
insertable.sort()
num_insertable = len(insertable)

selection = 0
# Set the initial selection to sand.
for i in range(num_insertable):
    if insertable[i][0] == 'sand':
        selection = i
        break

modify_radius = 5
old_pos = None

(ul_x, ul_y) = (0.0,0.0)
scale = 1.0
paused = False
pointing_at = None

def rebound():
    global ul_x, ul_y
    ul_x = min(1.0-scale, max(0, ul_x))
    ul_y = min(1.0-scale, max(0, ul_y))

def advance():
    global beet_state
    beet_state = beet.step(beet_state)

while True:
    start = time.time()
    if not paused:
        advance()
    frame = beet.render(beet_state, ul_x, ul_y, scale, width, height).get()
    end = time.time()
    futhark_time = (end-start)*1000

    start = time.time()
    pygame.surfarray.blit_array(screen, frame)
    end = time.time()
    blit_time = (end-start)*1000

    (mouse_x,mouse_y) = pygame.mouse.get_pos()
    if 0 <= mouse_x < width and 0 <= mouse_y < height:
        start = time.time()
        now_pointing_at = beet.element_at(beet_state,
                                          ul_x, ul_y, scale, width, height,
                                          mouse_x, mouse_y)
        if now_pointing_at != pointing_at:
            pointing_at = now_pointing_at
            pointing_at_name = element_name(now_pointing_at)
        end = time.time()
        query_time = (end-start)*1000
    else:
        query_time = 0

    speedmessage = "Stepping and rendering took %.2fms; blitting %.2fms; querying element name %.2fms" % \
                   (futhark_time, blit_time, query_time)

    showText(speedmessage, (10, 10))
    showText('%s (radius %d)' % (insertable[selection][0], modify_radius), (10,40))
    showText('Under cursor: %s' % pointing_at_name, (10,70))

    pygame.display.flip()

    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            sys.exit()
        elif event.type == pygame.MOUSEBUTTONDOWN:
            if event.button == 4:
                modify_radius = min(modify_radius + 1, 100)
            if event.button == 5:
                modify_radius = max(modify_radius - 1, 1)
        if event.type == pygame.KEYDOWN:
            if event.key == pygame.K_PAGEDOWN:
                selection = (selection + 1) % num_insertable
            elif event.key == pygame.K_PAGEUP:
                selection = (selection - 1) % num_insertable
            elif event.key == pygame.K_LEFT:
                ul_x -= 0.01
                rebound()
            elif event.key == pygame.K_RIGHT:
                ul_x += 0.01
                rebound()
            elif event.key == pygame.K_UP:
                ul_y -= 0.01
                rebound()
            elif event.key == pygame.K_DOWN:
                ul_y += 0.01
                rebound()
            elif event.unicode == '+':
                scale = max(1.0/min(width,height), scale*0.99)
                rebound()
            elif event.unicode == '-':
                scale = min(1.0, scale * 1.01)
                rebound()
            elif event.key == pygame.K_HOME:
                scale = 1.0
                (ul_x, ul_y) = (0.0, 0.0)
            elif event.key == pygame.K_SPACE:
                paused = True
                advance()
            elif event.key == pygame.K_RETURN:
                paused = False

    if pygame.mouse.get_pressed()[0] and pygame.mouse.get_pos() != None:
        # insert the selected element here.

        new_pos = pygame.mouse.get_pos()
        if old_pos == None:
            old_pos = new_pos

        args = [beet_state] + [ul_x, ul_y, scale, width, height] + list(old_pos) + list(new_pos) + [modify_radius, insertable[selection][1]]
        beet_state = beet.add_element(*args)
        old_pos = new_pos

    elif pygame.mouse.get_pressed()[2] and pygame.mouse.get_pos() != None:
        # remove any element element here.

        new_pos = pygame.mouse.get_pos()
        if old_pos == None:
            old_pos = new_pos

        args = [beet_state] + [ul_x, ul_y, scale, width, height] + list(old_pos) + list(new_pos) + [modify_radius]
        beet_state = beet.clear_element(*args)
        old_pos = new_pos
    else:
        old_pos = None

File: diving-beet.go
package main

import (
	"fmt"
	"github.com/veandco/go-sdl2/sdl"
	"github.com/veandco/go-sdl2/ttf"
	"sort"
	"time"
)

// ByName implements sort.Interface for []Element based on
// the Name field.
type ByName []Element

func (a ByName) Len() int           { return len(a) }
func (a ByName) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByName) Less(i, j int) bool { return a[i].Name < a[j].Name }

func min(x, y float64) float64 {
	if x < y {
		return x
	} else {
		return y
	}
}

func max(x, y float64) float64 {
	if x < y {
		return y
	} else {
		return x
	}
}

func main() {
	var err error

	screenX := 1024
	screenY := 768
	ul_x := 0.0
	ul_y := 0.0
	scale := 1.0
	radius := 10
	selected_element := 0
	mouse_down := false
	last_mouse_x := int32(0)
	last_mouse_y := int32(0)
	paused := false
	steps_per_frame := 10

	game := NewGame(screenX, screenY)
	defer game.Free()

	elements := game.Elements()
	sort.Sort(ByName(elements))

	if err := sdl.Init(sdl.INIT_EVERYTHING); err != nil {
		panic(err)
	}
	defer sdl.Quit()

	if err := ttf.Init(); err != nil {
		panic(err)
	}

	var font *ttf.Font
	text_size := 20
	if font, err = ttf.OpenFont("font.ttf", text_size); err != nil {
		panic(err)
	}
	defer font.Close()

	window, err := sdl.CreateWindow("test", sdl.WINDOWPOS_UNDEFINED, sdl.WINDOWPOS_UNDEFINED,
		int32(screenX), int32(screenY), sdl.WINDOW_SHOWN)
	if err != nil {
		panic(err)
	}
	defer window.Destroy()

	window_surface, err := window.GetSurface()
	if err != nil {
		panic(err)
	}

	frame_surface, err :=
		sdl.CreateRGBSurfaceFrom(game.Frame, int32(screenX), int32(screenY), 32, screenX*4, 0xFF0000, 0xFF00, 0xFF, 0x00000000)
	if err != nil {
		panic(err)
	}
	defer frame_surface.Free()

	white := sdl.Color{R: 255, G: 255, B: 255, A: 255}
	showText := func(s string, x, y int32) {
		var solid *sdl.Surface
		if solid, err = font.RenderUTF8Solid(s, white); err != nil {
			panic(err)
		}
		defer solid.Free()

		r := sdl.Rect{X: x, Y: y, W: 0, H: 0}
		if err = solid.Blit(nil, window_surface, &r); err != nil {
			panic(err)
		}
	}

	render := func() {
		start := time.Now()

		if !paused {
			for i := 0; i < steps_per_frame; i++ {
				game.Step()
			}
		}
		game.Render(ul_x, ul_y, scale, screenX, screenY)

		fut_time := time.Now().Sub(start)

		start = time.Now()
		if err = frame_surface.Blit(nil, window_surface, nil); err != nil {
			panic(err)
		}
		blit_time := time.Now().Sub(start)

		showText(
			fmt.Sprintf(
				"Futhark call took %.2fms; blitting took %.2fms.",
				fut_time.Seconds()*1000, blit_time.Seconds()*1000),
			0, 0)
		showText(
			fmt.Sprintf(
				"Inserting %s (radius %d)",
				elements[selected_element].Name, radius),
			0, int32(text_size+5))
		showText(
			fmt.Sprintf(
				"Under cursor %s", game.ElementAt(ul_x, ul_y, scale, last_mouse_x, last_mouse_y).Name),
			0, int32((text_size+5)*2))

		window.UpdateSurface()
	}

	rebound := func() {
		ul_x = min(1.0-scale, max(0, ul_x))
		ul_y = min(1.0-scale, max(0, ul_y))
	}

	onKeyboard := func(t sdl.KeyboardEvent) {
		if t.Type == sdl.KEYDOWN {
			switch t.Keysym.Sym {
			case sdl.K_RIGHT:
				ul_x += 0.01
				rebound()
			case sdl.K_LEFT:
				ul_x -= 0.01
				rebound()
			case sdl.K_UP:
				ul_y -= 0.01
				rebound()
			case sdl.K_DOWN:
				ul_y += 0.01
				rebound()
			case sdl.K_PLUS, sdl.K_p:
				scale = max(1/min(float64(screenX), float64(screenY)), scale*0.99)
				rebound()
			case sdl.K_MINUS, sdl.K_l:
				scale = min(1.0, scale*1.01)
				rebound()
			case sdl.K_SPACE:
				paused = true
				game.Step()
			case sdl.K_RETURN:
				paused = false

			case sdl.K_PAGEUP:
				selected_element = (selected_element + 1) % len(elements)
			case sdl.K_PAGEDOWN:
				selected_element = selected_element - 1
				if selected_element < 0 {
					selected_element = len(elements) - 1
				}
			}
		}
	}

	onMouseButton := func(t sdl.MouseButtonEvent) {
	}

	onMouseWheel := func(t sdl.MouseWheelEvent) {
		radius += int(t.Y)
		if radius < 1 {
			radius = 1
		}
	}

	checkMouse := func() {
		x, y, s := sdl.GetMouseState()
		if !mouse_down && s != 0 {
			mouse_down = true
			last_mouse_x = x
			last_mouse_y = y
		}

		if mouse_down {
			if s == 0 {
				mouse_down = false
			} else if (s & sdl.ButtonLMask()) != 0 {
				game.AddElem(ul_x, ul_y, scale, last_mouse_x, last_mouse_y, x, y, radius, elements[selected_element])
			} else if (s & sdl.ButtonRMask()) != 0 {
				game.ClearElem(ul_x, ul_y, scale, last_mouse_x, last_mouse_y, x, y, radius)
			}
		}

		last_mouse_x, last_mouse_y = x, y
	}

	running := true
	for running {
		render()

		checkMouse()

		for event := sdl.PollEvent(); event != nil; event = sdl.PollEvent() {
			switch t := event.(type) {
			case *sdl.QuitEvent:
				println("Quit")
				running = false
			case *sdl.KeyboardEvent:
				onKeyboard(*t)
			case *sdl.MouseButtonEvent:
				onMouseButton(*t)
			case *sdl.MouseWheelEvent:
				onMouseWheel(*t)
			}
		}
	}
}
